### PR TITLE
chore: auth on api-v2 with api-key

### PR DESCRIPTION
**Pull request by ThyMinimalDev**
## What does this PR do?

Enables api calls to api-v2 to use api-key as authentication token

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [x] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

yarn test api-auth.strategy.spec.ts

### CODE REVIEW

**Comment by ThyMinimalDev** [comment]
e2e and self review coming soon

File: apps/api/v2/src/modules/auth/guards/permissions/permissions.guard.ts [hunk 1]
```diff
@@ -17,13 +22,20 @@ export class PermissionsGuard implements CanActivate {
...
       return false;
     }
 
-    const oAuthClientPermissions = await this.getOAuthClientPermissions(accessToken);
+    const isApiKey = authString?.startsWith(this.config.get("api.apiKeyPrefix") ?? "cal_");
+
+    // only check permissions for accessTokens attached to an oAuth Client
+    if (isApiKey) {
+      return true;
+    }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/auth/guards/permissions/permissions.guard.ts:36]: apiKey based auth does not care about oauth client permissions

File: apps/api/v2/src/modules/auth/guards/permissions/permissions.guard.spec.ts [hunk 2]
```diff
@@ -13,7 +14,7 @@ describe("PermissionsGuard", () => {
 
   beforeEach(async () => {
     reflector = new Reflector();
-    guard = new PermissionsGuard(reflector, createMock<TokensRepository>());
+    guard = new PermissionsGuard(reflector, createMock<TokensRepository>(), createMock<ConfigService>());
```
Comment (sean-brydon) [inline apps/api/v2/src/modules/auth/guards/permissions/permissions.guard.spec.ts:17]: NIT: 

Why not test and mock this the way nestjs suggestes. 

https://docs.nestjs.com/fundamentals/testing#testing-utilities 
e.g. 
``` 
  let guard: PermissionGuard;
  let tokenRepository: TokensRepository;
  let configService: ConfigService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        PermissionGuard,
        {
          provide: TokensRepository,
          useValue: {
              /* jest fns can go here */
          },
        },
        {
          provide: ConfigService,
          useValue: {
            get: jest.fn(),
          },
        },
      ],
    }).compile();

    guard = module.get<PermissionGuard>(PermissionGuard);
    tokenRepository = module.get<TokensRepository>(TokensRepository);
    configService = module.get<ConfigService>(ConfigService);
  });
```

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  const validApiKeyEmail = "api-key-user-email@example.com";
+  const validAccessTokenEmail = "access-token-user-email@example.com";
+  let validApiKeyUser: User;
+  let validAccessTokenUser: User;
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        ApiAuthStrategy,
+        {
```
Comment (sean-brydon) [inline apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.spec.ts:43]: NIT: I had a case like this in private API and extracted this to a ConfigServiceMock util something like. 

```ts
class ConfigServiceMock {
  private readonly config = {
    "db.readUrl": "postgresql://postgres:@localhost:5450/calendso",
    "db.writeUrl": "postgresql://postgres:@localhost:5450/calendso",
    "db.redisUrl": "redis://localhost:6379",
    "api.keyPrefix": "cal_test_",
    "e2e": true,
    "api.licenseKey": "license-key",
    "api.licenseKeyUrl": "license-key-url.cal.com",
    "env": "development"
  };

  get(key: string): any {
    return this.config[key] ?? null;
  }
}
```


Then the testing module would become 

```ts
   {
        provide: ConfigService,
        useClass: ConfigServiceMock,
      }
        ```

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (!ownerId) {
+      throw new UnauthorizedException(INVALID_ACCESS_TOKEN);
+    }
+
+    const user: UserWithProfile | null = await this.userRepository.findByIdWithProfile(ownerId);
+    return user;
+  }
+
+  async authenticate(request: Request) {
```
Comment (supalarry) [inline apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts:82]: NIT: can we move this function to top of service so it's easier to read code top down.

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    private readonly tokensRepository: TokensRepository,
+    private readonly userRepository: UsersRepository,
+    private readonly apiKeyRepository: ApiKeyRepository
+  ) {
+    super();
+  }
+
+  async apiKeyStrategy(apiKey: string) {
+    const isLicenseValid = await this.deploymentsService.checkLicense();
+    if (!isLicenseValid && this.config.get("env") === "production") {
```
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts:30]: true, we should have a license key set even in dev

[the remark this answers is not in the document]
Comment (sean-brydon) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts:30]: Keep in mind - license key in dev will require you to have console/private-api running at all times. 

(We warn when no license key is provided in the check in calcom/web in dev but just fail when trying to use when env === prod)

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.spec.ts (hunk 3 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.spec.ts:43]: the main reason we need to do this here is that we dont create a Nest App like we do in e2e specs, meaning that the env variables are not loading in the config service which is normally done in main.ts

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts:30]: in dev we can just add a license key to the deployments table in the database no ?

[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts:30]: which is the same behaviour than in prod

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.spec.ts (hunk 3 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.spec.ts:43]: found a better way to write it

File: apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/auth/strategies/api-auth/api-auth.strategy.ts]: confirmed adding a license key in the dev database and placing the same key in env variables reproduce what prod does

**Review by supalarry — DISMISSED** [review]

**Comment by ThyMinimalDev** [comment]
Ready for merge

**Review by supalarry — APPROVED** [review]

**Pull request merged**